tox.ini

```diff
@@ -27,6 +29,43 @@ passenv =
   HTTPS_PROXY
   NO_PROXY
 
+[testenv:deploy-{amd64,arm64}]
+description = Deploy local charm with promql-transform
+allowlist_externals =
+    wget
+    tar
+    juju
+commands =
+    # Download arch-appropriate promql-transform
+    amd64: wget -O {temp_dir}/promql-transform.tar.gz {[vars]promql_amd64_archive}
+    arm64: wget -O {temp_dir}/promql-transform.tar.gz {[vars]promql_arm64_archive}
+    tar --overwrite -C {temp_dir} -xzvf {temp_dir}/promql-transform.tar.gz promql-transform
+    amd64: juju deploy --resource prometheus-image=ubuntu/prometheus:latest --resource promql-transform-amd64={temp_dir}/promql-transform ./prometheus-k8s_ubuntu-20.04-amd64.charm {posargs}
+    arm64: juju deploy --resource prometheus-image=ubuntu/prometheus:latest --resource promql-transform-arm64={temp_dir}/promql-transform ./prometheus-k8s_ubuntu-20.04-arm64.charm {posargs}
+
+[testenv:refresh-{amd64,arm64}]
+description = refresh charm from local path, with promql-transform
+allowlist_externals =
+    wget
+    tar
+    juju
+commands =
+    # Download arch-appropriate promql-transform
+    amd64: wget -O {temp_dir}/promql-transform.tar.gz {[vars]promql_amd64_archive}
+    arm64: wget -O {temp_dir}/promql-transform.tar.gz {[vars]promql_arm64_archive}
+    tar --overwrite -C {temp_dir} -xzvf promql-transform.tar.gz promql-transform
+    amd64: juju refresh --resource promql-transform-amd64={temp_dir}/promql-transform --path=./prometheus-k8s_ubuntu-20.04-amd64.charm {posargs}
+    arm64: juju refresh --resource promql-transform-arm64={temp_dir}/promql-transform --path=./prometheus-k8s_ubuntu-20.04-arm64.charm {posargs}
+
+[testenv:publish-{edge,beta}]
+description = release a previously uploaded charm to a charmhub edge/beta channel
+allowlist_externals =
+    charmcraft
+commands =
+    # Usage example: tox -e publish-edge -- --revision=19
+    edge: charmcraft release prometheus-k8s --channel=edge --resource=prometheus-image:1 --resource=promql-transform-amd64:3 {posargs}
```

Comment:
Wouldn't both resource revisions need to be customizable ?

Reply:
Correct, but it's a tradeoff between:
1. resources don't get updated often so might as well hard code them until they need to change
2. make it customizable but require users to run `tox -e publish-edge -- --revision=19 --resource=prometheus-image:1 --resource=promql-transform-amd64:3`

(2) is almost the same length as than the original command though.

Reply:
Hmm. We could end up with the same problem then. We publish new resources and forget to updated the tox.ini so we end up making releases with old resources.

Reply:
Correct, but this is slightly better than omitting.
I am fine with dropping the publish tox env, just wanted to show how it would look like.
It has pluses and minuses obviously.

Reply:
True. But not my call to make.

Reply:
The GH action to publish charm automatically fetches whatever is specified anyway.

We noticed initially because every merge to Grafana ends up putting Grafana 8 (latest) on `edge`, so this is probably less of a concern anyway. The action may be useful for manually deploying akin to a `make` target if there's an emergency fix which needs to go out while code review is still pending, though.

Reply:
I sort of think we should just not have this. All the options should be configurable at which point why not just run charmcraft?

Reply:
> why not just run charmcraft?

Most charms only have "image" resource, but some charms have also file resource, which are forgettable.

> All the options should be configurable

This tox env suggests the opposite: most of the time, most options are fixed.

Reply:
> The GH action to publish charm automatically fetches whatever is specified anyway.
> 
> We noticed initially because every merge to Grafana ends up putting Grafana 8 (latest) on `edge`, so this is probably less of a concern anyway. The action may be useful for manually deploying akin to a `make` target if there's an emergency fix which needs to go out while code review is still pending, though.

Is what you're saying that it does not respect pinning done in upstream-source? iIf so, you should file a bug in the actions repo.

Reply:
Ping @rbarry82

Reply:
We've had some time away from this, and... considering the amount of "this adds risks/non-deterministic behavior to make the lives of our own team marginally easier" things we've come across, I don't think it's a great idea.

We're now hardcoding "fixed" options to specific versions. What will we do when/if `promql-transform` bumps and we may have a difference in stable/edge and/or some other stream? What about having a different ROCK in edge versus stable? What if someone forgets to use this because they are relying on the GH action and the image "goes backwards" do one which has a security vulnerability?

* I will not reinvent `make` badly with `tox`.
* I will not reinvent `make` badly with `tox`.
* I will not reinvent `make` badly with `tox`.
* I will not reinvent `make` badly with `tox`.
* I will not reinvent `make` badly with `tox`.

Considering that `tox` is not built for (and is terrible at) any kind of dynamic/tagged updating, a GH action which can be triggered (with or without [selectable inputs ala Jenkins](https://github.community/t/how-to-pass-input-parameters-to-github-actions/123941)), or which is able to set these in a more flexible way.

We should not recreate `make` with far less flexibility/functionality. Hyrum's Law applies here. Such a tox env (deploy/refresh) will pretty much never be used by anyone other than people writing patches for/testing this charm, and `juju deploy` is reasonably good at error logging anyway, telling authors "you didn't give me a value for `--resource foo`"). 

If we're concerned about publishing and forgetting, we should open an issue with `charmcraft` (and they can push it back to `charmrepo | charmstore | juju/charmhub` if it's better handled there). 

It is, simply, just not the right tool for the job. Any job, really. Publishing to edge should be done through CI, and stable/beta can _also_ be done through CI with parameterized workflows when we get there. Those workflows can pick out "appropriate" values without hardcoding them or `sh -c '...'` (from putting a square `tox` in a round "automation" hole), and `tox -e deploy-{arch} || tox -e refresh-{arch} ...` are better off just being shell aliases, as they're purely developer workflows.

Then they could even use `dotenv` or similar to discover this

Add them to a `zsh`